From b9864a0e0d848f85156a4893df79f7ad8ac01258 Mon Sep 17 00:00:00 2001
From: Evlers <1425295900@qq.com>
Date: Thu, 6 Apr 2023 14:50:24 +0800
Subject: [PATCH 2/2] Modify the device tree and readme file

Add mangopi_mq_quad_defconfig file
Add reg_vcc3v3 node to mmc0
Disable ehci0 and ohci0 node
Modify dr_mode is otg for usbotg node
Delete the parameters greater than 1.512GHz from the cpu_opp_table
Modify the armbian README.md file
---
 README.md                                     | 197 ++++--------------
 .../dts/allwinner/sun50i-h616-cpu-opp.dtsi    |  30 +--
 .../allwinner/sun50i-h616-mangopi-mcore.dts   |  30 ++-
 .../arm64/configs/mangopi_mq_quad_defconfig   |   2 +-
 4 files changed, 78 insertions(+), 181 deletions(-)
 rename default_linux-sunxi64-edge.config => arch/arm64/configs/mangopi_mq_quad_defconfig (99%)

diff --git a/README.md b/README.md
index 74c4c16416d3..5e6db2370405 100644
--- a/README.md
+++ b/README.md
@@ -1,154 +1,43 @@
-Mainline linux kernel for Orange Pi PC/PC2/PC3/One, TBS A711, PinePhone (Pro), PocketBook Touch Lux 3
-------------------------------------------------------------------------------------------------------
-
-This kernel tree is meant for:
-
-- Orange Pi One
-- Orange Pi PC
-- Orange Pi PC 2
-- Orange Pi 3
-- PinePhone 1.0, 1.1 and 1.2(a/b)
-- TBS A711 Tablet
-- PocketBook Touch Lux 3
-- Pinebook Pro
-- Pinephone Pro
-
-Features in addition to mainline:
-
-- [Orange Pi One/PC/PC2] More aggressive OPPs for CPU
-- [All] Mark one of DRM planes as a cursor plane, speeding up Xorg based desktop with modesetting driver
-- [Orange Pi One/PC/PC2] Configure on-board micro-switches to perform system power off function
-- [Orange Pi One/PC/PC2/3] HDMI audio
-- [Orange Pi 3] Ethernet
-- [TBS A711] HM5065 (back camera) / GC2145 (front camera)
-- [PinePhone] WiFi, Bluetooth, Audio, Modem power, HDMI out over USB-C, USB-C support, cameras, PMIC improvements, power management, fixes here and there
-- [PocketBook Touch Lux 3] Display and Touchscreen support
-- [Pinephone Pro] Everything
-
-Pre-built u-boot and kernels are available at https://xff.cz/kernels/
-
-You may need some firmware files for some part of the functionality. Those are
-available at: https://megous.com/git/linux-firmware
-
-If you want to reproduce my pre-built kernels exactly, you'll need to uncomment
-CONFIG_EXTRA_FIRMWARE_DIR and CONFIG_EXTRA_FIRMWARE in the defconfigs, and
-point CONFIG_EXTRA_FIRMWARE_DIR to a directory on your computer where the
-clone of https://megous.com/git/linux-firmware resides.
-
-You can also leave those two config options commented out, and copy the contents
-of https://megous.com/git/linux-firmware to /lib/firmware/ on the target device.
-
-You can use this kernel to run a desktop environment on Orange Pi SBCs,
-Arch Linux on your Pinephone, or to have a completely opensource OS on
-a Pocketbook e-ink book reader.
-
-Have fun!
-
-
-Build instructions
-------------------
-
-These are rudimentary instructions and you need to understand what you're doing.
-These are just core steps required to build the ATF/u-boot/kernel. Downloading,
-verifying, renaming to correct directories is not described or mentioned. You
-should be able to infer missing necessary steps yourself for your particular needs.
-
-Get necessary toolchains from:
-
-- https://releases.linaro.org/components/toolchain/binaries/latest/aarch64-linux-gnu/ for 64bit Orange Pi PC2 and Orange Pi 3, PinePhone
-- https://releases.linaro.org/components/toolchain/binaries/latest/arm-linux-gnueabihf/ for 32bit Orange Pis, Pocketbook, TBS tablet
-
-Extract toolchains and prepare the environment:
-
-    CWD=`pwd`
-    OUT=$CWD/builds
-    SRC=$CWD/u-boot
-    export PATH="$PATH:$CWD/Toolchains/arm/bin:$CWD/Toolchains/aarch64/bin"
-
-For Orange Pi PC2, Orange Pi 3 or PinePhone:
-
-    export CROSS_COMPILE=aarch64-linux-gnu-
-    export KBUILD_OUTPUT=$OUT/.tmp/uboot-pc2
-    rm -rf "$KBUILD_OUTPUT"
-    mkdir -p $KBUILD_OUTPUT $OUT/pc2
-
-Get and build ATF from https://github.com/ARM-software/arm-trusted-firmware:
-
-    make -C "$CWD/arm-trusted-firmware" PLAT=sun50i_a64 DEBUG=1 bl31
-    cp "$CWD/arm-trusted-firmware/build/sun50i_a64/debug/bl31.bin" "$KBUILD_OUTPUT"
-
-Use sun50i_a64 for Orange Pi PC2 or PinePhone and sun50i_h6 for Orange Pi 3.
-
-Build u-boot from https://megous.com/git/u-boot/ (opi-v2020.04 branch) with appropriate
-defconfig (orangepi_one_defconfig, orangepi_pc2_defconfig, orangepi_pc_defconfig, orangepi_3_defconfig, tbs_a711_defconfig, pinephone_defconfig).
-
-My u-boot branch already has all the necessary patches integrated and is configured for quick u-boot/kernel startup.
-
-    make -C u-boot orangepi_pc2_defconfig
-    make -C u-boot -j5
-    
-    cp $KBUILD_OUTPUT/.config $OUT/pc2/uboot.config
-    cat $KBUILD_OUTPUT/{spl/sunxi-spl.bin,u-boot.itb} > $OUT/pc2/uboot.bin
-
-Get kernel from this repository and checkout the latest orange-pi-5.18 branch.
-
-Build the kernel for 64-bit boards:
-
-    export ARCH=arm64
-    export CROSS_COMPILE=aarch64-linux-gnu-
-    export KBUILD_OUTPUT=$OUT/.tmp/linux-arm64
-    mkdir -p $KBUILD_OUTPUT $OUT/pc2
-
-    make -C linux orangepi_defconfig
-    # or make -C linux pocketbook_touch_lux_3_defconfig
-    # or make -C linux tbs_a711_defconfig
-    make -C linux -j5 clean
-    make -C linux -j5 Image dtbs
-
-    cp -f $KBUILD_OUTPUT/arch/arm64/boot/Image $OUT/pc2/
-    cp -f $KBUILD_OUTPUT/.config $OUT/pc2/linux.config
-    cp -f $KBUILD_OUTPUT/arch/arm64/boot/dts/allwinner/sun50i-h5-orangepi-pc2.dtb $OUT/pc2/board.dtb
-
-Build the kernel for 32-bit boards:
-
-    export ARCH=arm
-    export CROSS_COMPILE=arm-linux-gnueabihf-
-    export KBUILD_OUTPUT=$OUT/.tmp/linux-arm
-    mkdir -p $KBUILD_OUTPUT $OUT/pc
-
-    make orangepi_defconfig
-    # or make pinephone_defconfig
-    make -C linux orangepi_defconfig
-    make -C linux -j5 clean
-    make -C linux -j5 zImage dtbs
-    
-    cp -f $KBUILD_OUTPUT/arch/arm/boot/zImage $OUT/pc/
-    cp -f $KBUILD_OUTPUT/.config $OUT/pc/linux.config
-    cp -f $KBUILD_OUTPUT/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dtb $OUT/pc/board.dtb
-    # Or use sun8i-h3-orangepi-one.dtb for Orange Pi One
-
-
-PinePhone
----------
-
-I don't run u-boot on PinePhone, so my pre-built kernel packages don't come
-with u-boot built for PinePhone.
-
-
-Kernel lockup issues
---------------------
-
-*If you're getting lockups on boot or later during thermal regulation,
-you're missing an u-boot patch.*
-
-This patch is necessary to run this kernel!
-
-These lockups are caused by improper NKMP clock factors selection
-in u-boot for PLL_CPUX. (M divider should not be used. P divider
-should be used only for frequencies below 240MHz.)
-
-This patch for u-boot fixes it:
-
-  0001-sunxi-h3-Fix-PLL1-setup-to-never-use-dividers.patch
-
-Kernel side is already fixed in this kernel tree.
+## MangoPI MQ-Quad board linux base on Allwinner H616
+Linux-5.19.4
+<br>
+
+## Peripheral support
+| **Peripheral** |               **Describe**               |
+|----------------|------------------------------------------|
+|   RTL8723DS    | WiFi/BT                                  |
+|   USB Host     | 1 typec host with2 usb host in misc port |
+|   USB OTG      | 1 typec otg                              |
+|   ethernet     | ethernet for mac1 with ephy in misc port |
+|     UART       | uart1 with uart2                         |
+|  Mini HDMI     | hdmi with audio                          |
+|      I2C       | i2c1 with i2c2                           |
+|  Nor Flash     | spi0 nor flash                           |
+|      SPI       | spi1dev drivers                          |
+|      GPU       | gpu(panforst)                            |
+
+## Build
+```
+# using mangopi_mq_quad_defconfig
+make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- mangopi_mq_quad_defconfig
+
+# build kernel image
+make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -jx Image
+# output file: arch/arm64/boot/Image
+
+# build device tree
+make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -jx dtbs
+# output file: arch/arm64/boot/dts/allwinner/sun50i-h616-mangopi-mcore.dtb
+
+# build modules and install modules
+make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -jx modules
+make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- INSTALL_MOD_PATH=<any-path-you-like> modules modules_install
+# After a successful build, you can find the module in the provided INSTALL_MOD_PATH directory
+
+# install headers files
+make ARCH=arm64 INSTALL_HDR_PATH=<any-path-you-like> headers_install
+# Change INSTALL_HDR_PATH to the path you want to install to, such as /usr or <some_prefix>/usr to facilitate late cross-compilation
+```
+
+## Reference
+https://github.com/open-cores/mangguo-h616-armbian
diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi
index 495d79740a50..02c41ed2230b 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi
@@ -70,23 +70,23 @@ opp-1512000000 {
 		        opp-microvolt = <1120000 1120000 1200000>;
 		};
 		
-		opp-1608000000 {
-			clock-latency-ns = <244144>; /* 8 32k periods */
-			opp-hz = /bits/ 64 <1608000000>;
-			opp-microvolt = <1160000 1160000 1200000>;
-		};
+		// opp-1608000000 {
+		// 	clock-latency-ns = <244144>; /* 8 32k periods */
+		// 	opp-hz = /bits/ 64 <1608000000>;
+		// 	opp-microvolt = <1160000 1160000 1200000>;
+		// };
 
-		opp-1704000000 {
-			clock-latency-ns = <244144>; /* 8 32k periods */
-			opp-hz = /bits/ 64 <1704000000>;
-			opp-microvolt = <1180000 1180000 1200000>;
-		};
+		// opp-1704000000 {
+		// 	clock-latency-ns = <244144>; /* 8 32k periods */
+		// 	opp-hz = /bits/ 64 <1704000000>;
+		// 	opp-microvolt = <1180000 1180000 1200000>;
+		// };
 
-		opp-1800000000 {
-			clock-latency-ns = <244144>; /* 8 32k periods */
-			opp-hz = /bits/ 64 <1800000000>;
-			opp-microvolt = <1200000 1200000 1200000>;
-		};
+		// opp-1800000000 {
+		// 	clock-latency-ns = <244144>; /* 8 32k periods */
+		// 	opp-hz = /bits/ 64 <1800000000>;
+		// 	opp-microvolt = <1200000 1200000 1200000>;
+		// };
 												
 	};
 };
diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h616-mangopi-mcore.dts b/arch/arm64/boot/dts/allwinner/sun50i-h616-mangopi-mcore.dts
index 8140417fb04b..a49df71b5c37 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h616-mangopi-mcore.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h616-mangopi-mcore.dts
@@ -62,6 +62,14 @@ reg_vcc5v: vcc5v {
 		regulator-always-on;
 	};
 
+	reg_vcc3v3: vcc3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
+	};
+
 	reg_usb1_vbus: usb1-vbus {
 		compatible = "regulator-fixed";
 		regulator-name = "usb1-vbus";
@@ -84,7 +92,7 @@ &gpu {
 };
 
 &ehci0 {
-	status = "okay";
+	status = "disabled";
 };
 
 &ehci1 {
@@ -139,7 +147,7 @@ emd_rgmii_phy: ethernet-phy@0 {
 
 
 &mmc0 {
-	vmmc-supply = <&reg_aldo1>;
+	vmmc-supply = <&reg_vcc3v3>;
 	cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>;	/* PF6 */
 	bus-width = <4>;
 	status = "okay";
@@ -190,7 +198,7 @@ spidev@0 {
 };
 
 &ohci0 {
-	status = "okay";
+	status = "disabled";
 };
 
 &ohci1 {
@@ -283,17 +291,17 @@ &uart0 {
 };
 
 &uart1 {
- uart-has-rtscts;
- pinctrl-0 = <&uart1_pins>, <&uart1_rts_cts_pins>;
- pinctrl-names = "default";
- status = "okay";
+	uart-has-rtscts;
+	pinctrl-0 = <&uart1_pins>, <&uart1_rts_cts_pins>;
+	pinctrl-names = "default";
+	status = "okay";
 
-   bluetooth {
+	bluetooth {
               compatible = "realtek,rtl8723ds-bt";
               device-wake-gpios = <&pio 6 17 GPIO_ACTIVE_HIGH>; /* PG17 */
               enable-gpios = <&pio 6 19 GPIO_ACTIVE_HIGH>; /* PG19 */
               host-wake-gpios = <&pio 6 16 GPIO_ACTIVE_HIGH>; /* PG16 */
- };
+	};
 };
 
 &uart2 {
@@ -303,8 +311,8 @@ &uart2 {
 };
 
 &usbotg {
-	dr_mode = "host";	/* USB A type receptable */
-        status = "okay";
+	dr_mode = "otg";	/* USB A type receptable */
+	status = "okay";
 };
 
 &usbphy {
diff --git a/default_linux-sunxi64-edge.config b/arch/arm64/configs/mangopi_mq_quad_defconfig
similarity index 99%
rename from default_linux-sunxi64-edge.config
rename to arch/arm64/configs/mangopi_mq_quad_defconfig
index 90c67f48e3b6..a82e658230e8 100644
--- a/default_linux-sunxi64-edge.config
+++ b/arch/arm64/configs/mangopi_mq_quad_defconfig
@@ -2819,7 +2819,7 @@ CONFIG_RTL8723DS=m
 CONFIG_RTL8822CS=m
 # CONFIG_RTL8822BU is not set
 CONFIG_RTL8821CU=m
-CONFIG_88XXAU=m
+# CONFIG_88XXAU is not set
 CONFIG_RTL8192EU=m
 CONFIG_RTL8189FS=m
 CONFIG_RTL8189ES=m
-- 
2.34.1

